docs/site/Interceptors.md

```diff
@@ -0,0 +1,529 @@
+---
+lang: en
+title: 'Interceptors'
+keywords: LoopBack 4.0, LoopBack 4
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/Interceptors.html
+---
+
+## Overview
+
+Interceptors are reusable functions to provide aspect-oriented logic around
+method invocations. There are many use cases for interceptors, such as:
+
+- Add extra logic before / after method invocation, for example, logging or
+  measuring method invocations.
+- Validate/transform arguments
+- Validate/transform return values
+- Catch/transform errors, for example, normalize error objects
+- Override the method invocation, for example, return from cache
+
+## Basic use
+
+### Interceptors on controllers
+
+Interceptors are supported for public controller methods (including both static
+and prototype) and handler functions for REST routes.
+
+Controller methods decorated with `@intercept` are invoked with applied
+interceptors for corresponding routes upon API requests.
+
+```ts
+import {intercept} from '@loopback/context';
+
+@intercept(log)
+export class OrderController {
+  @intercept('caching-interceptor')
+  async listOrders(userId: string) {
+    // ...
+  }
+}
+```
+
+It's also possible to configure global interceptors that are invoked before
+method level interceptors. For example, the following code registers a global
+`caching-interceptor` for all methods.
+
+```ts
+app
+  .bind('caching-interceptor')
+  .toProvider(CachingInterceptorProvider)
+  .apply(asInterceptor);
+```
+
+Global interceptors are also executed for route handler functions without a
+controller class. See an example in
+[Route Handler](Routes.md#using-partial-openapi-spec-fragments).
+
+### Create a proxy to apply interceptors
+
+A
+[proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy)
+can be created from the target class or object to apply interceptors. This is
+useful for the case that a controller declares dependencies of repositories or
+services and would like to allow repository or service methods to be
+intercepted.
```

Comment:
While I see how compelling this idea is, I find it also problematic. Consider the situation when a sync function is decorated with an async interceptor. The proxy is changing the API of the decorated class, turning synchronous methods into asynchronous.

Reply:
See https://github.com/strongloop/loopback-next/pull/2687/commits/a419b4fe47d3db95112fae84d596652b87450711

Reply:
> See a419b4f

@raymondfeng the commit does not seem to be a part of this pull request anymore, perhaps after some rebase & force-push changes? I left few comments for you in that commit, please take a look.

Reply:
This is the new commit to address your concerns - https://github.com/strongloop/loopback-next/pull/2687/commits/7b15940b66372c6cf042cd3a8bcf7a80a4d1caa9